flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/EpochManager.java

```diff
@@ -0,0 +1,202 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.asyncprocessing;
+
+import javax.annotation.Nullable;
+
+import java.util.LinkedList;
+
+/**
+ * Epoch manager segments inputs into distinct epochs, marked by the arrival of non-records(e.g.
+ * watermark, record attributes). Records are assigned to a unique epoch based on their arrival,
+ * records within an epoch are allowed to be parallelized, while the non-record of an epoch can only
+ * be executed when all records in this epoch have finished.
+ *
+ * <p>For more details please refer to FLIP-425.
+ */
+public class EpochManager {
+
+    /**
+     * This enum defines whether parallel execution between epochs is allowed. We should keep this
+     * internal and away from API module for now, until we could see the concrete need for {@link
+     * #PARALLEL_BETWEEN_EPOCH} from average users.
+     */
+    public enum ParallelMode {
+        /**
+         * Subsequent epochs must wait until the previous epoch is completed before they can start.
+         */
+        SERIAL_BETWEEN_EPOCH,
+        /**
+         * Subsequent epochs can begin execution even if the previous epoch has not yet completed.
+         * Usually performs better than {@link #SERIAL_BETWEEN_EPOCH}.
+         */
+        PARALLEL_BETWEEN_EPOCH
+    }
+
+    /**
+     * The reference to the {@link AsyncExecutionController}, used for {@link
+     * ParallelMode#SERIAL_BETWEEN_EPOCH}. Can be null when testing.
+     */
+    final AsyncExecutionController asyncExecutionController;
+
+    /** The number of epochs that have arrived. */
+    long epochNum;
+
+    /** The output queue to hold ongoing epochs. */
+    LinkedList<Epoch> outputQueue;
+
+    /** Current active epoch, only one active epoch at the same time. */
+    Epoch activeEpoch;
+
+    public EpochManager(AsyncExecutionController aec) {
+        this.epochNum = 0;
+        this.outputQueue = new LinkedList<>();
+        // preset an empty epoch, the epoch action will be updated when non-record is received.
+        this.activeEpoch = new Epoch(epochNum++);
+        this.outputQueue.add(activeEpoch);
+        this.asyncExecutionController = aec;
+    }
+
+    /**
+     * Add a record to the current epoch and return the current open epoch, the epoch will be
+     * associated with the {@link RecordContext} of this record. Must be invoked within task thread.
+     *
+     * @return the current open epoch.
+     */
+    public Epoch onRecord() {
+        activeEpoch.ongoingRecordCount++;
+        return activeEpoch;
+    }
+
+    /**
+     * Add a non-record to the current epoch, close current epoch and open a new epoch. Must be
+     * invoked within task thread.
+     *
+     * @param action the action associated with this non-record.
+     * @param parallelMode the parallel mode for this epoch.
+     */
+    public void onNonRecord(Runnable action, ParallelMode parallelMode) {
```

Comment:
How about changing Runnable to ThrowingRunnable? This can help reduce duplicated try-catch code blocks in invokers.

Reply:
Because there is no exception in the signature of `disposeContext()` , if we change Runnable to ThrowingRunnable here, we need to catch the exception in `aec.disposeContext()` or  `epochManager.completeOneRecord()`.

I think that catching exceptions in `StreamOperator` is more convenient

Reply:
Got it. Given that we agreed to add `AsyncExecutionController.processNonRecord` in another comment and that `AsyncExecutionController` has `exceptionHandler`, how about keeping EpochManager using Runnable, and change `AsyncExecutionController.processNonRecord` to use ThrowingRunnable?